### PR TITLE
Integrated `conversion-verifier` in CI

### LINKED ISSUE #3085: Conversion is not completely implemented for all API types
/kind bug

Hey folks,
in core CAPI we have a tool called conversion-verifier which loops over all types and ensures the controller-runtime conversion interfaces are implemented correctly, i.e. it checks the types have the Hub / ConvertTo / ConvertFrom functions.

The results for CAPA are:
* type "infrastructure.cluster.x-k8s.io/AWSClusterTemplateList" in package "sigs.k8s.io/cluster-api-provider-aws/api/v1beta1" marked as storage version but it's not convertible, missing Hub() method
* type "infrastructure.cluster.x-k8s.io/AWSClusterTemplateList" in package "sigs.k8s.io/cluster-api-provider-aws/api/v1alpha4" it's not convertible, missing ConvertFrom() and ConvertTo() methods

I think it would be great to:
* implement the missing methods
* integrate the conversion-verifier in CI

The conversion-verifier can be found here:
* Code: https://github.com/kubernetes-sigs/cluster-api/tree/main/hack/tools/conversion-verifier
* Usage: https://github.com/kubernetes-sigs/cluster-api/blob/main/Makefile#L339-L341 (core CAPI runs it as part of the verify job)

I can provide additional details if needed / if you want to implement this issue.

**Environment:**

- Cluster-api-provider-aws version: latest
- Kubernetes version: (use `kubectl version`): -
- OS (e.g. from `/etc/os-release`): -

### LINKED ISSUE #3127: Add conversion-verifier to CI
Populating this issue from https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/3085 to track conversion-verifier addition separately.

```
Integrate the conversion-verifier in CI.

The conversion-verifier can be found here:

Code: https://github.com/kubernetes-sigs/cluster-api/tree/main/hack/tools/conversion-verifier
Usage: https://github.com/kubernetes-sigs/cluster-api/blob/main/Makefile#L339-L341 (core CAPI runs it as part of the verify job)

```

/good-first-issue
/priority backlog

### PR DESCRIPTION
**Pull request by zeborg**
Signed-off-by: Abhinav Sinha <abhinav@nirmata.com>

**What type of PR is this?**
`/kind feature`

**What this PR does / why we need it**:
* Added conversion functions for `AWSClusterTemplateList`
* Added integration for `conversion-verifier` in CI with `verify-conversions` make target
* Added k8s boilerplate text to `cmd/clusterawsadm/ami/helper_test.go`

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3085, fixes #3127 

**Special notes for your reviewer**:

**Checklist**:

- [x] squashed commits
- [ ] includes documentation
- [ ] adds unit tests
- [ ] adds or updates e2e tests

**Release note**:

```release-note
Integrated `conversion-verifier` in CI
```

### CODE REVIEW

File: hack/tools/Makefile [hunk 1]
```diff
@@ -76,6 +76,10 @@ CONVERSION_GEN := $(BIN_DIR)/conversion-gen
 $(CONVERSION_GEN): $(BIN_DIR) go.mod go.sum $(call rwildcard,.,third_party/conversion-gen/*.*)
 	go build -tags=tools -o $@ ./third_party/conversion-gen
 
+CONVERSION_VERIFIER := $(BIN_DIR)/conversion-verifier
+$(CONVERSION_VERIFIER): $(BIN_DIR) go.mod go.sum
+	go build -tags=tools -o $@ ./third_party/conversion-verifier
```
Comment (richardcase) [inline hack/tools/Makefile:81]: I think we should build it from the cluster-api repo instead of mnainting our own version of the tool.

Comment (richardcase) [reply hack/tools/Makefile:81]: https://github.com/kubernetes-sigs/cluster-api/tree/main/hack/tools/conversion-verifier

**Review by richardcase — CHANGES_REQUESTED** [review]
I think we should avoid having our own copy of this utility. Would you be able to look at changing it so that we use the code from CAPI?

File: hack/tools/Makefile (hunk 1 above)
Comment (zeborg) [reply hack/tools/Makefile:81]: Yes, you're right, I'll fix this asap. Thanks for pointing it out!

Comment (richardcase) [reply hack/tools/Makefile]: Thanks

**Review by richardcase — COMMENTED** [review]
One small comment. Also could you squash your commits?

File: Makefile [hunk 2]
```diff
@@ -604,6 +605,10 @@ verify-modules: modules
 		echo "go module files are out of date"; exit 1; \
 	fi
 
+.PHONY: verify-conversions
+verify-conversions: $(CONVERSION_VERIFIER)  ## Verifies expected API conversion are in place
+	$(CONVERSION_VERIFIER)
```
Comment (richardcase) [inline Makefile:610]: Could you add an echo as part of this target so that we can see its running?

Comment (zeborg) [reply Makefile:611]: Done 👍

**Comment by zeborg** [comment]
Hey @richardcase, I've squashed the commits as you suggested earlier.

**Review by richardcase — APPROVED** [review]
Thanks for making those changes

/lgtm

**Comment by richardcase** [comment]
Assigning for final approval

/assign sedefsavas

**Pull request merged**